dashboard/app/models/course_version.rb

```diff
@@ -18,4 +18,31 @@
 
 class CourseVersion < ApplicationRecord
   belongs_to :content_root, polymorphic: true
+
+  def self.update_course_version(content_root)
+    # TODO: Once the Course model is added, change this to just be version_year, since then the
+    # unique index will be on (course_id, key).
+    key = "#{content_root.family_name}-#{content_root.version_year}"
+
+    unless content_root.is_course?
+      CourseVersion.find_by(key: key)&.destroy
+      content_root.course_version = nil
+      content_root.save
+      return nil
+    end
+
+    course_version = CourseVersion.find_or_create_by(
+      key: key,
+      display_name: content_root.version_year,
+      content_root: content_root,
+    )
+
+    # Delete old CourseVersion if the key has been changed to something else
+    content_root.course_version.destroy if course_version != content_root.course_version
+    content_root.course_version = course_version
```

Comment:
If you just do this does it no delete the old course version?

Reply:
It does not, as far as I can tell when I was testing this out.

Reply:
Here's what I get when I try that. I think what happens is ActiveRecord just tries to set the CourseVersion object's content_root to nil. So it tries to unassociate it rather than delete it. Not sure if there's some setting or some other approach for getting it to delete instead.

```
[1] pry(CourseVersion)> content_root
=> #<Script:0x00007fc0d62287a8
 id: 56,
 name: "test-all-properties",
 created_at: Tue, 11 Aug 2020 05:02:28 UTC +00:00,
 updated_at: Tue, 11 Aug 2020 05:02:28 UTC +00:00,
 wrapup_video_id: nil,
 hidden: true,
 user_id: nil,
 login_required: false,
 properties: {},
 new_name: nil,
 family_name: nil>
[2] pry(CourseVersion)> content_root.course_version
=> #<CourseVersion:0x00007fc112679960
 id: 7,
 key: "-2020",
 display_name: "2020",
 properties: nil,
 content_root_type: "Script",
 content_root_id: 56,
 created_at: Tue, 11 Aug 2020 05:02:28 UTC +00:00,
 updated_at: Tue, 11 Aug 2020 05:02:28 UTC +00:00>
[3] pry(CourseVersion)> content_root.course_version = nil
ActiveRecord::StatementInvalid: Mysql2::Error: Column 'content_root_id' cannot be null: UPDATE `course_versions` SET `content_root_id` = NULL, `updated_at` = '2020-08-11 05:02:48' WHERE `course_versions`.`id` = 7
```

Reply:
Good question Dani, I was curious about this too. It's not explicitly spelled out in the docs, but it's possible that the `dependent: :destroy` option is what makes this work the way you're describing in other situations. Here are the parts of the docs which seem relevant (this is a has_one and is also polymorphic)

* https://guides.rubyonrails.org/association_basics.html#options-for-has-one-dependent These docs mention destroying the associated object when the parent is deleted, but does not explicitly say the same will happen upon assignment. 
* https://guides.rubyonrails.org/association_basics.html#polymorphic-associations doesn't mention destroying so not very helpful

By contrast, the docs for for has_many ... `dependent: :destroy` are more explicit:

* https://guides.rubyonrails.org/association_basics.html#dependent

    > The :destroy and :delete_all options also affect the semantics of the collection.delete and collection= methods by causing them to destroy associated objects when they are removed from the collection.